src/functions.php

```diff
@@ -758,27 +758,34 @@ function _parse_message($message)
         throw new \InvalidArgumentException('Invalid message');
     }
 
-    // Iterate over each line in the message, accounting for line endings
-    $lines = preg_split('/(\\r?\\n)/', $message, -1, PREG_SPLIT_DELIM_CAPTURE);
-    $result = ['start-line' => array_shift($lines), 'headers' => [], 'body' => ''];
-    array_shift($lines);
-
-    for ($i = 0, $totalLines = count($lines); $i < $totalLines; $i += 2) {
-        $line = $lines[$i];
-        // If two line breaks were encountered, then this is the end of body
-        if (empty($line)) {
-            if ($i < $totalLines - 1) {
-                $result['body'] = implode('', array_slice($lines, $i + 2));
-            }
-            break;
-        }
-        if (strpos($line, ':')) {
-            $parts = explode(':', $line, 2);
-            $key = trim($parts[0]);
-            $value = isset($parts[1]) ? trim($parts[1]) : '';
-            $result['headers'][$key][] = $value;
-        }
-    }
+    $headerDelimiterPosition = strpos($message, "\r\n\r\n");
```

Comment:
Any leading `\r\n` should be stripped here and it needs to check whether `$headerDelimiterPosition` isn't `false`. See also [here](https://github.com/amphp/http-server/blob/e215ff364be22614f9d22331c36d66056ddb8330/src/Driver/Http1Driver.php#L205-L211).

Reply:
Ah, nice catch.